java/code/src/com/redhat/rhn/frontend/xmlrpc/system/SystemHandler.java

```diff
@@ -4569,6 +4614,57 @@ public int schedulePackageRemoveByNevra(User loggedInUser, final Integer sid, Li
                 ActionFactory.TYPE_PACKAGES_REMOVE, allowModules)[0].intValue();
     }
 
+    /**
+     * Schedule package lock for a system.
+     *
+     * @param loggedInUser The current user
+     * @param sid ID of the server
+     * @param pkgIdsToLock List of package IDs to lock (as Integers)
+     * @param pkgIdsToUnlock List of package IDs to lock (as Integers)
+     * @param earliestOccurrence Earliest occurrence of the package install
+     * @return package action id
+     *
+     * @xmlrpc.doc Schedule package installation for a system.
+     * @xmlrpc.param #param("string", "sessionKey")
+     * @xmlrpc.param #param("int", "serverId")
+     * @xmlrpc.param #array_single("int", "pkgIdsToLock")
+     * @xmlrpc.param #array_single("int", "pkgIdsToUnlock")
+     * @xmlrpc.param dateTime.iso8601 earliestOccurrence
+     * @xmlrpc.returntype #param_desc("return_int_success")
+     */
+    public Long schedulePackageLockChange(User loggedInUser, Long sid,
+                                          List<Long> pkgIdsToLock,
+                                          List<Long> pkgIdsToUnlock, Date earliestOccurrence) {
+
+        Server server = SystemManager.lookupByIdAndUser(sid.longValue(), loggedInUser);
+        DataResult<PackageListItem> lockedPackagesResult =
+                PackageManager.systemLockedPackages(server.getId(), null);
+
+        Set<Package> pkgsAlreadyLocked = lockedPackagesResult.stream().map(
+                pkg -> PackageManager.lookupByIdAndUser(pkg.getPackageId(), loggedInUser)).collect(Collectors.toSet());
+        Set<Package> pkgsToLock = pkgIdsToLock.stream().map(
+                pkg -> PackageManager.lookupByIdAndUser(pkg.longValue(), loggedInUser)).
+                filter(Objects::isNull).collect(Collectors.toSet());
+        pkgsToLock.stream().forEach(pkg -> pkg.setLockPending(Boolean.TRUE));
+
+        Set<Package> pkgsToUnlock = pkgIdsToUnlock.stream().map(
+                        x -> PackageManager.lookupByIdAndUser(x.longValue(), loggedInUser)).
+                filter(Objects::isNull).collect(Collectors.toSet());
+
+        PackageManager.setPendingStatusOnLockedPackages(pkgsToUnlock, PackageManager.PKG_PENDING_UNLOCK);
+
+        PackageManager.lockPackages(server.getId(), pkgsToLock);
+
+        pkgsToLock.addAll(pkgsAlreadyLocked);
+        try {
+            Action a = ActionManager.schedulePackageLock(loggedInUser, pkgsToLock, earliestOccurrence, server);
+            return a.getId();
+
+        }
+        catch (com.redhat.rhn.taskomatic.TaskomaticApiException e) {
+            throw new TaskomaticApiException(e.getMessage());
```

Comment:
Why is this re-throw needed? It seems to me that it would be the same if we don't catch the exception. Or should we maybe log something?

Reply:
Although I agree, I followed the pattern presents in the other methods of this class...my feeling is it's something related to xmlrpc implementation.. @mcalmer what do you think?

Reply:
Oh, this is not obvious. These are 2 different Exceptions with the same name but in different Packages.
Don't ask me why we have 2.

The new one is "com.redhat.rhn.frontend.**xmlrpc**.TaskomaticApiException" which is a "Fault" Exceptions special for XMLRPC.